distros/M/Math-GSL/pm/Math/GSL/Permutation.pm.2.3
# This file was automatically generated by SWIG (http://www.swig.org).
# Version 4.0.1
#
# Do not make changes to this file unless you know what you are doing--modify
# the SWIG interface file instead.

package Math::GSL::Permutation;
use base qw(Exporter);
use base qw(DynaLoader);
package Math::GSL::Permutationc;
bootstrap Math::GSL::Permutation;
package Math::GSL::Permutation;
@EXPORT = qw();

# ---------- BASE METHODS -------------

package Math::GSL::Permutation;

sub TIEHASH {
    my ($classname,$obj) = @_;
    return bless $obj, $classname;
}

sub CLEAR { }

sub FIRSTKEY { }

sub NEXTKEY { }

sub FETCH {
    my ($self,$field) = @_;
    my $member_func = "swig_${field}_get";
    $self->$member_func();
}

sub STORE {
    my ($self,$field,$newval) = @_;
    my $member_func = "swig_${field}_set";
    $self->$member_func($newval);
}

sub this {
    my $ptr = shift;
    return tied(%$ptr);
}


# ------- FUNCTION WRAPPERS --------

package Math::GSL::Permutation;

*gsl_error = *Math::GSL::Permutationc::gsl_error;
*gsl_stream_printf = *Math::GSL::Permutationc::gsl_stream_printf;
*gsl_strerror = *Math::GSL::Permutationc::gsl_strerror;
*gsl_set_error_handler = *Math::GSL::Permutationc::gsl_set_error_handler;
*gsl_set_error_handler_off = *Math::GSL::Permutationc::gsl_set_error_handler_off;
*gsl_set_stream_handler = *Math::GSL::Permutationc::gsl_set_stream_handler;
*gsl_set_stream = *Math::GSL::Permutationc::gsl_set_stream;
*gsl_permute_complex_long_double = *Math::GSL::Permutationc::gsl_permute_complex_long_double;
*gsl_permute_complex_long_double_inverse = *Math::GSL::Permutationc::gsl_permute_complex_long_double_inverse;
*gsl_permute_complex = *Math::GSL::Permutationc::gsl_permute_complex;
*gsl_permute_complex_inverse = *Math::GSL::Permutationc::gsl_permute_complex_inverse;
*gsl_permute_complex_float = *Math::GSL::Permutationc::gsl_permute_complex_float;
*gsl_permute_complex_float_inverse = *Math::GSL::Permutationc::gsl_permute_complex_float_inverse;
*gsl_permute_long_double = *Math::GSL::Permutationc::gsl_permute_long_double;
*gsl_permute_long_double_inverse = *Math::GSL::Permutationc::gsl_permute_long_double_inverse;
*gsl_permute = *Math::GSL::Permutationc::gsl_permute;
*gsl_permute_inverse = *Math::GSL::Permutationc::gsl_permute_inverse;
*gsl_permute_int = *Math::GSL::Permutationc::gsl_permute_int;
*gsl_permute_int_inverse = *Math::GSL::Permutationc::gsl_permute_int_inverse;
*gsl_permute_uint = *Math::GSL::Permutationc::gsl_permute_uint;
*gsl_permute_uint_inverse = *Math::GSL::Permutationc::gsl_permute_uint_inverse;
*gsl_permute_vector = *Math::GSL::Permutationc::gsl_permute_vector;
*gsl_permute_vector_inverse = *Math::GSL::Permutationc::gsl_permute_vector_inverse;
*gsl_permute_vector_int = *Math::GSL::Permutationc::gsl_permute_vector_int;
*gsl_permute_vector_int_inverse = *Math::GSL::Permutationc::gsl_permute_vector_int_inverse;
*gsl_permute_vector_uint = *Math::GSL::Permutationc::gsl_permute_vector_uint;
*gsl_permute_vector_uint_inverse = *Math::GSL::Permutationc::gsl_permute_vector_uint_inverse;
*gsl_permute_vector_long = *Math::GSL::Permutationc::gsl_permute_vector_long;
*gsl_permute_vector_long_inverse = *Math::GSL::Permutationc::gsl_permute_vector_long_inverse;
*gsl_permute_vector_long_double = *Math::GSL::Permutationc::gsl_permute_vector_long_double;
*gsl_permute_vector_long_double_inverse = *Math::GSL::Permutationc::gsl_permute_vector_long_double_inverse;
*gsl_permute_vector_ulong = *Math::GSL::Permutationc::gsl_permute_vector_ulong;
*gsl_permute_vector_ulong_inverse = *Math::GSL::Permutationc::gsl_permute_vector_ulong_inverse;
*gsl_permute_vector_short = *Math::GSL::Permutationc::gsl_permute_vector_short;
*gsl_permute_vector_short_inverse = *Math::GSL::Permutationc::gsl_permute_vector_short_inverse;
*gsl_permute_vector_ushort = *Math::GSL::Permutationc::gsl_permute_vector_ushort;
*gsl_permute_vector_ushort_inverse = *Math::GSL::Permutationc::gsl_permute_vector_ushort_inverse;
*gsl_permutation_alloc = *Math::GSL::Permutationc::gsl_permutation_alloc;
*gsl_permutation_calloc = *Math::GSL::Permutationc::gsl_permutation_calloc;
*gsl_permutation_init = *Math::GSL::Permutationc::gsl_permutation_init;
*gsl_permutation_free = *Math::GSL::Permutationc::gsl_permutation_free;
*gsl_permutation_memcpy = *Math::GSL::Permutationc::gsl_permutation_memcpy;
*gsl_permutation_fread = *Math::GSL::Permutationc::gsl_permutation_fread;
*gsl_permutation_fwrite = *Math::GSL::Permutationc::gsl_permutation_fwrite;
*gsl_permutation_fscanf = *Math::GSL::Permutationc::gsl_permutation_fscanf;
*gsl_permutation_fprintf = *Math::GSL::Permutationc::gsl_permutation_fprintf;
*gsl_permutation_size = *Math::GSL::Permutationc::gsl_permutation_size;
*gsl_permutation_data = *Math::GSL::Permutationc::gsl_permutation_data;
*gsl_permutation_swap = *Math::GSL::Permutationc::gsl_permutation_swap;
*gsl_permutation_valid = *Math::GSL::Permutationc::gsl_permutation_valid;
*gsl_permutation_reverse = *Math::GSL::Permutationc::gsl_permutation_reverse;
*gsl_permutation_inverse = *Math::GSL::Permutationc::gsl_permutation_inverse;
*gsl_permutation_next = *Math::GSL::Permutationc::gsl_permutation_next;
*gsl_permutation_prev = *Math::GSL::Permutationc::gsl_permutation_prev;
*gsl_permutation_mul = *Math::GSL::Permutationc::gsl_permutation_mul;
*gsl_permutation_linear_to_canonical = *Math::GSL::Permutationc::gsl_permutation_linear_to_canonical;
*gsl_permutation_canonical_to_linear = *Math::GSL::Permutationc::gsl_permutation_canonical_to_linear;
*gsl_permutation_inversions = *Math::GSL::Permutationc::gsl_permutation_inversions;
*gsl_permutation_linear_cycles = *Math::GSL::Permutationc::gsl_permutation_linear_cycles;
*gsl_permutation_canonical_cycles = *Math::GSL::Permutationc::gsl_permutation_canonical_cycles;
*gsl_permutation_get = *Math::GSL::Permutationc::gsl_permutation_get;

############# Class : Math::GSL::Permutation::gsl_permutation_struct ##############

package Math::GSL::Permutation::gsl_permutation_struct;
use vars qw(@ISA %OWNER %ITERATORS %BLESSEDMEMBERS);
@ISA = qw( Math::GSL::Permutation );
%OWNER = ();
%ITERATORS = ();
*swig_size_get = *Math::GSL::Permutationc::gsl_permutation_struct_size_get;
*swig_size_set = *Math::GSL::Permutationc::gsl_permutation_struct_size_set;
*swig_data_get = *Math::GSL::Permutationc::gsl_permutation_struct_data_get;
*swig_data_set = *Math::GSL::Permutationc::gsl_permutation_struct_data_set;
sub new {
    my $pkg = shift;
    my $self = Math::GSL::Permutationc::new_gsl_permutation_struct(@_);
    bless $self, $pkg if defined($self);
}

sub DESTROY {
    return unless $_[0]->isa('HASH');
    my $self = tied(%{$_[0]});
    return unless defined $self;
    delete $ITERATORS{$self};
    if (exists $OWNER{$self}) {
        Math::GSL::Permutationc::delete_gsl_permutation_struct($self);
        delete $OWNER{$self};
    }
}

sub DISOWN {
    my $self = shift;
    my $ptr = tied(%$self);
    delete $OWNER{$ptr};
}

sub ACQUIRE {
    my $self = shift;
    my $ptr = tied(%$self);
    $OWNER{$ptr} = 1;
}


# ------- VARIABLE STUBS --------

package Math::GSL::Permutation;

*GSL_VERSION = *Math::GSL::Permutationc::GSL_VERSION;
*GSL_MAJOR_VERSION = *Math::GSL::Permutationc::GSL_MAJOR_VERSION;
*GSL_MINOR_VERSION = *Math::GSL::Permutationc::GSL_MINOR_VERSION;
*GSL_POSZERO = *Math::GSL::Permutationc::GSL_POSZERO;
*GSL_NEGZERO = *Math::GSL::Permutationc::GSL_NEGZERO;
*GSL_SUCCESS = *Math::GSL::Permutationc::GSL_SUCCESS;
*GSL_FAILURE = *Math::GSL::Permutationc::GSL_FAILURE;
*GSL_CONTINUE = *Math::GSL::Permutationc::GSL_CONTINUE;
*GSL_EDOM = *Math::GSL::Permutationc::GSL_EDOM;
*GSL_ERANGE = *Math::GSL::Permutationc::GSL_ERANGE;
*GSL_EFAULT = *Math::GSL::Permutationc::GSL_EFAULT;
*GSL_EINVAL = *Math::GSL::Permutationc::GSL_EINVAL;
*GSL_EFAILED = *Math::GSL::Permutationc::GSL_EFAILED;
*GSL_EFACTOR = *Math::GSL::Permutationc::GSL_EFACTOR;
*GSL_ESANITY = *Math::GSL::Permutationc::GSL_ESANITY;
*GSL_ENOMEM = *Math::GSL::Permutationc::GSL_ENOMEM;
*GSL_EBADFUNC = *Math::GSL::Permutationc::GSL_EBADFUNC;
*GSL_ERUNAWAY = *Math::GSL::Permutationc::GSL_ERUNAWAY;
*GSL_EMAXITER = *Math::GSL::Permutationc::GSL_EMAXITER;
*GSL_EZERODIV = *Math::GSL::Permutationc::GSL_EZERODIV;
*GSL_EBADTOL = *Math::GSL::Permutationc::GSL_EBADTOL;
*GSL_ETOL = *Math::GSL::Permutationc::GSL_ETOL;
*GSL_EUNDRFLW = *Math::GSL::Permutationc::GSL_EUNDRFLW;
*GSL_EOVRFLW = *Math::GSL::Permutationc::GSL_EOVRFLW;
*GSL_ELOSS = *Math::GSL::Permutationc::GSL_ELOSS;
*GSL_EROUND = *Math::GSL::Permutationc::GSL_EROUND;
*GSL_EBADLEN = *Math::GSL::Permutationc::GSL_EBADLEN;
*GSL_ENOTSQR = *Math::GSL::Permutationc::GSL_ENOTSQR;
*GSL_ESING = *Math::GSL::Permutationc::GSL_ESING;
*GSL_EDIVERGE = *Math::GSL::Permutationc::GSL_EDIVERGE;
*GSL_EUNSUP = *Math::GSL::Permutationc::GSL_EUNSUP;
*GSL_EUNIMPL = *Math::GSL::Permutationc::GSL_EUNIMPL;
*GSL_ECACHE = *Math::GSL::Permutationc::GSL_ECACHE;
*GSL_ETABLE = *Math::GSL::Permutationc::GSL_ETABLE;
*GSL_ENOPROG = *Math::GSL::Permutationc::GSL_ENOPROG;
*GSL_ENOPROGJ = *Math::GSL::Permutationc::GSL_ENOPROGJ;
*GSL_ETOLF = *Math::GSL::Permutationc::GSL_ETOLF;
*GSL_ETOLX = *Math::GSL::Permutationc::GSL_ETOLX;
*GSL_ETOLG = *Math::GSL::Permutationc::GSL_ETOLG;
*GSL_EOF = *Math::GSL::Permutationc::GSL_EOF;

@EXPORT_OK = qw/
                gsl_permutation_alloc
                gsl_permutation_calloc
                gsl_permutation_init
                gsl_permutation_free
                gsl_permutation_memcpy
                gsl_permutation_fread
                gsl_permutation_fwrite
                gsl_permutation_fscanf
                gsl_permutation_fprintf
                gsl_permutation_size
                gsl_permutation_data
                gsl_permutation_get
                gsl_permutation_swap
                gsl_permutation_valid
                gsl_permutation_reverse
                gsl_permutation_inverse
                gsl_permutation_next
                gsl_permutation_prev
                gsl_permutation_mul
                gsl_permutation_linear_to_canonical
                gsl_permutation_canonical_to_linear
                gsl_permutation_inversions
                gsl_permutation_linear_cycles
                gsl_permutation_canonical_cycles
                gsl_permute
                gsl_permute_inverse
                gsl_permute_int
                gsl_permute_int_inverse
                gsl_permute_vector
                gsl_permute_vector_inverse
                gsl_permute_vector_int
                gsl_permute_vector_int_inverse
            /;
%EXPORT_TAGS = ( all => [ @EXPORT_OK ] );

### wrapper interface ###
sub new {
    my ($class, $value) = @_;
    my $this = {};
    $this->{_length} = $value;
    $this->{_permutation} = gsl_permutation_calloc($value);
    bless $this, $class;
}

sub as_list {
    my $self=shift;
    $self->get( [ 0 .. $self->length - 1  ] );
}

sub get {
    my ($self, $indices) = @_;
    return  map {  gsl_permutation_get($self->{_permutation}, $_ ) } @$indices ;
}

sub raw { (shift)->{_permutation} }
sub length { (shift)->{_length} }




__END__

=encoding utf8

=head1 NAME

Math::GSL::Permutation - functions for creating and manipulating permutations

=head1 SYNOPSIS

 use Math::GSL::Permutation qw/:all/;
 my $permutation = Math::GSL::Permutation->new(30); # allocate and initialize a permutation of size 30
 my $lenght = $permutation->lenght; # returns the lenght of the permutation object, here it is 30
 gsl_permutation_swap($permutation->raw, 2,7);
 # the raw method is made to use the underlying permutation structure of the permutation object
 my $value = $permutation->get(2); # returns the third value (starting from 0) of the permutation
 my @values = $permutation->as_list; # returns all the values of the permutation
 my @set = $permutation->get([0,1,2,3]); # returns the four first values of the permutation

=head1 DESCRIPTION

Here is a list of all the functions included in this module :

=over

=item gsl_permutation_alloc($n) - return a newly allocated permutation of size $n

=item gsl_permutation_calloc($n) - return a newly allocated permutation of size $n which is initialized to the identity

=item gsl_permutation_init($p) - initialize the permutation $p to the identity i.e. (0,1,2, ..., n-1)

=item gsl_permutation_free($p) - free all the memory use by the permutaion $p

=item gsl_permutation_memcpy($dest, $src) - copy the permutation $src into the permutation $dest, the two permutations must have the same lenght and return 0 if the operation suceeded, 1 otherwise

=item gsl_permutation_fread($stream, $p) -  This function reads into the permutation $p from the open stream $stream (opened with the gsl_fopen function from the Math::GSL module) in binary format. The permutation $p must be preallocated with the correct length since the function uses the size of $p to determine how many bytes to read. The function returns 1 if there was a problem reading from the file. The data is assumed to have been written in the native binary format on the same architecture.

=item gsl_permutation_fwrite($stream, $p) - This function writes the elements of the permutation $p to the stream $stream (opened with the gsl_fopen function from the Math::GSL module) in binary format. The function returns 1 if there was a problem writing to the file. Since the data is written in the native binary format it may not be portable between different architectures.

=item gsl_permutation_fscanf($stream, $p) - This function reads formatted data from the stream $stream (opened with the gsl_fopen function from the Math::GSL module) into the permutation $p. The permutation $p must be preallocated with the correct length since the function uses the size of $p to determine how many numbers to read. The function returns 1 if there was a problem reading from the file.

=item gsl_permutation_fprintf($stream, $p, $format) - This function writes the elements of the permutation $p line-by-line to the stream $stream (opened with the gsl_fopen function from the Math::GSL module) using the format specifier $format, which should be suitable. "%zu\n" is a suitable format. The function returns 1 if there was a problem writing to the file.

=item gsl_permutation_size($p) - return the size of the permutation $p

=item gsl_permutation_data

=item gsl_permutation_get($p, $i) - return the $i-th element of the permutation $p, return 0 if $i is outside the range of 0 to n-1

=item gsl_permutation_swap($p, $i, $j) - exchange the $i-th position and the $j-th position of the permutation $p and return 0 if the operation suceeded, 1 otherwise

=item gsl_permutation_valid($p) - return 0 if the permutation $p is valid (if the n elements contain each of the numbers 0 to n-1 once and only once), 1 otherwise

=item gsl_permutation_reverse($p) - reverse the elements of the permutation $p

=item gsl_permutation_inverse($inv, $p) - compute the inverse of the permutation $p, storing it in $inv and return 0 if the operation succeeded, 1 otherwise

=item gsl_permutation_next($p) - advance the permutation $p to the next permutation in lexicographic order and return 0 if the operation succeeded, 1 otherwise

=item gsl_permutation_prev($p) - step backward from the permutation $p to the previous permutation in lexicographic order and return 0 if the operation suceeded, 1 otherwise

=item gsl_permutation_mul($p, $pa, $pb) - combine the two permutation $pa and $pb into a single permutation $p and return 0 if the operation suceeded, 1 otherwise

=item gsl_permutation_linear_to_canonical($q, $p) - compute the canonical form the permutation $p and store it in $q and return 0 if the operation suceeded, 1 otherwise

=item gsl_permutation_canonical_to_linear($p, $q) - convert a canonical permutation $q back into linear form and store it in $p and return 0 if the operation suceeded, 1 otherwise

=item gsl_permutation_inversions($p) - return the number of inversions in the permutation $p

=item gsl_permutation_linear_cycles($p) - return the number of cycles in the permutation $p, given a linear form

=item gsl_permute_vector_int_inversegsl_permutation_canonical_cycles($p) - return the number of cycles in the permutation $p, given a canonical form

=item gsl_permute

=item gsl_permute_inverse

=item gsl_permute_int

=item gsl_permute_int_inverse

=item gsl_permute_vector

=item gsl_permute_vector_inverse

=item gsl_permute_vector_int

=back

 You have to add the functions you want to use inside the qw/put_funtion_here/ with spaces between each function.
 You can also write use Math::GSL::CDF qw/:all/ to use all avaible functions of the module.
 Other tags are also avaible, here is a complete list of all tags for this module.
For more informations on the functions, we refer you to the GSL offcial documentation:
L<http://www.gnu.org/software/gsl/manual/html_node/>



=head1 EXAMPLES

 use Math::GSL::Permutation qw/:all/;
 $p->{permutation} = gsl_permutation_calloc(5);
 print "The permutation contains [";
 map { print gsl_permutation_get($p->{permutation}, $_) . ", " } (0..3);
 print gsl_permutation_get($p->{permutation}, 4);
 print "] \n";
 print "We'll then swap the first and last elements of the permutation...\n";
 gsl_permutation_swap($p->{permutation}, 0, 4);
 print "The permutation now contains [";
 map { print gsl_permutation_get($p->{permutation},$_) . ", " } (0..3);
 print gsl_permutation_get($p->{permutation}, 4);
 print "] \n";


 use Math::GSL::Permutation qw/:all/;
 use Math::GSL::Vector qw/:all/;
 my $p->{permutation} = gsl_permutation_calloc(6);
 gsl_permutation_init($p->{permutation});
 gsl_permutation_swap($p->{permutation}, 0, 1);
 print "The permutation has his first and second elements swapped : [";
 map { print gsl_permutation_get($p->{permutation}, $_) . "," } (0..4);
 print gsl_permutation_get($p->{permutation}, 5) . "] \n";
 my $vec->{vector} = gsl_vector_alloc(6);
 map { gsl_vector_set($vec->{vector}, $_, $_) } (0..5);

 print "We will now apply the permutation to this vector : [";
 map { print gsl_vector_get($vec->{vector}, $_) . "," } (0..4);
 print gsl_vector_get($vec->{vector}, 5) . "] \n";
 gsl_permute_vector($p->{permutation}, $vec->{vector});
 print "The vector is now : [";
 map { print gsl_vector_get($vec->{vector}, $_) . "," } (0..4);
 print gsl_vector_get($vec->{vector}, 5) . "] \n";


=head1 AUTHORS

Jonathan "Duke" Leto <jonathan@leto.net> and Thierry Moisan <thierry.moisan@gmail.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008-2020 Jonathan "Duke" Leto and Thierry Moisan

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
1;


distros/W/Wurst/src/wurstsrc/gsldir/permutation.c
/* permutation/permutation.c
 * 
 * Copyright (C) 1996, 1997, 1998, 1999, 2000 Brian Gough
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include "gsl_errno.h"
#include "gsl_permutation.h"

size_t gsl_permutation_size(const gsl_permutation * p)
{
    return p->size;
}

size_t *gsl_permutation_data(const gsl_permutation * p)
{
    return p->data;
}

#ifndef HIDE_INLINE_STATIC
size_t gsl_permutation_get(const gsl_permutation * p, const size_t i)
{
    if (gsl_check_range) {
	if (i >= p->size) {	/* size_t is unsigned, can't be negative */
	    GSL_ERROR_VAL("index out of range", GSL_EINVAL, 0);
	}
    }

    return p->data[i];
}
#endif

int
gsl_permutation_swap(gsl_permutation * p, const size_t i, const size_t j)
{
    const size_t size = p->size;

    if (i >= size) {
	GSL_ERROR("first index is out of range", GSL_EINVAL);
    }

    if (j >= size) {
	GSL_ERROR("second index is out of range", GSL_EINVAL);
    }

    if (i != j) {
	size_t tmp = p->data[i];
	p->data[i] = p->data[j];
	p->data[j] = tmp;
    }

    return GSL_SUCCESS;
}


int gsl_permutation_valid(gsl_permutation * p)
{
    const size_t size = p->size;

    size_t i, j;

    for (i = 0; i < size; i++) {
	if (p->data[i] >= size) {
	    GSL_ERROR("permutation index outside range", GSL_FAILURE);
	}

	for (j = 0; j < i; j++) {
	    if (p->data[i] == p->data[j]) {
		GSL_ERROR("duplicate permutation index", GSL_FAILURE);
	    }
	}
    }

    return GSL_SUCCESS;
}

void gsl_permutation_reverse(gsl_permutation * p)
{
    const size_t size = p->size;

    size_t i;

    for (i = 0; i < (size / 2); i++) {
	size_t j = size - i - 1;

	size_t tmp = p->data[i];
	p->data[i] = p->data[j];
	p->data[j] = tmp;
    }
}

int
gsl_permutation_inverse(gsl_permutation * inv, const gsl_permutation * p)
{
    const size_t size = p->size;

    size_t i;

    if (inv->size != size) {
	GSL_ERROR("permutation lengths are not equal", GSL_EBADLEN);
    }

    for (i = 0; i < size; i++) {
	inv->data[p->data[i]] = i;
    }

    return GSL_SUCCESS;
}

int gsl_permutation_next(gsl_permutation * p)
{
    /* Replaces p with the next permutation (in the standard lexicographical
     * ordering).  Returns GSL_FAILURE if there is no next permutation.
     */
    const size_t size = p->size;
    size_t i, j, k;

    if (size < 2) {
	return GSL_FAILURE;
    }

    i = size - 2;

    while ((p->data[i] > p->data[i + 1]) && (i != 0)) {
	i--;
    }

    if ((i == 0) && (p->data[0] > p->data[1])) {
	return GSL_FAILURE;
    }

    k = i + 1;

    for (j = i + 2; j < size; j++) {
	if ((p->data[j] > p->data[i]) && (p->data[j] < p->data[k])) {
	    k = j;
	}
    }

    /* swap i and k */

    {
	size_t tmp = p->data[i];
	p->data[i] = p->data[k];
	p->data[k] = tmp;
    }

    for (j = i + 1; j <= ((size + i) / 2); j++) {
	size_t tmp = p->data[j];
	p->data[j] = p->data[size + i - j];
	p->data[size + i - j] = tmp;
    }

    return GSL_SUCCESS;
}

int gsl_permutation_prev(gsl_permutation * p)
{
    const size_t size = p->size;
    size_t i, j, k;

    if (size < 2) {
	return GSL_FAILURE;
    }

    i = size - 2;

    while ((p->data[i] < p->data[i + 1]) && (i != 0)) {
	i--;
    }

    if ((i == 0) && (p->data[0] < p->data[1])) {
	return GSL_FAILURE;
    }

    k = i + 1;

    for (j = i + 2; j < size; j++) {
	if ((p->data[j] < p->data[i]) && (p->data[j] > p->data[k])) {
	    k = j;
	}
    }

    /* swap i and k */

    {
	size_t tmp = p->data[i];
	p->data[i] = p->data[k];
	p->data[k] = tmp;
    }

    for (j = i + 1; j <= ((size + i) / 2); j++) {
	size_t tmp = p->data[j];
	p->data[j] = p->data[size + i - j];
	p->data[size + i - j] = tmp;
    }

    return GSL_SUCCESS;
}

int
gsl_permutation_mul(gsl_permutation * p, const gsl_permutation * pa,
		    const gsl_permutation * pb)
{
    size_t i;
    const size_t size = p->size;

    if (pa->size != size) {
	GSL_ERROR("size of result does not match size of pa", GSL_EINVAL);
    }

    if (pb->size != size) {
	GSL_ERROR("size of result does not match size of pb", GSL_EINVAL);
    }

    for (i = 0; i < size; i++) {
	p->data[i] = pb->data[pa->data[i]];
    }

    return GSL_SUCCESS;
}
int
gsl_permutation_memcpy(gsl_permutation * dest, const gsl_permutation * src)
{
    const size_t src_size = src->size;
    const size_t dest_size = dest->size;

    if (src_size != dest_size) {
	GSL_ERROR("permutation lengths are not equal", GSL_EBADLEN);
    }

    {
	size_t j;

	for (j = 0; j < src_size; j++) {
	    dest->data[j] = src->data[j];
	}
    }

    return GSL_SUCCESS;
}
